presto-hive/src/main/java/com/facebook/presto/hive/HiveRecordSink.java

```diff
@@ -92,29 +103,170 @@
+            handle.getColumnTypes(),
+            handle.getColumnTypes(),
+            handle.getHiveStorageFormat().getOutputFormat(),
+            handle.getHiveStorageFormat().getSerDe(),
+            null,
+            "",
+            false,
+            handle.getConnectorSession());
+    }
+
+    private HiveRecordSink(Path target,
+                             JobConf conf,
+                             List<String> columnNames,
+                             List<String> dataColumnNames,
+                             List<Type> columnTypes,
+                             List<Type> dataColumnTypes,
+                             String outputFormat,
+                             String serdeLib,
+                             Map<String, String> serdeParameters,
+                             String filePrefix,
+                             boolean isPartitioned,
+                             ConnectorSession connectorSession)
+    {
+        checkNotNull(target, "Base path for table data not set");
+
+        basePath = target;
+        fieldCount = columnNames.size();
+        dataFieldsCount = dataColumnNames.size();
 
-        sampleWeightField = handle.getColumnNames().indexOf(SAMPLE_WEIGHT_COLUMN_NAME);
-        columnTypes = ImmutableList.copyOf(handle.getColumnTypes());
-        connectorSession = handle.getConnectorSession();
-        hasDateTimeTypes = handle.getColumnTypes().stream().map(this::containsDateTime).collect(toList());
+        this.conf = conf;
+        this.isPartitioned = isPartitioned;
 
-        Iterable<String> hiveTypeNames = transform(transform(handle.getColumnTypes(), HiveType::toHiveType), HiveType::getHiveTypeName);
+        sampleWeightField = columnNames.indexOf(SAMPLE_WEIGHT_COLUMN_NAME);
+        this.columnTypes = columnTypes;
+        this.connectorSession = connectorSession;
+        hasDateTimeTypes = columnTypes.stream().map(this::containsDateTime).collect(toList());
 
-        Properties properties = new Properties();
-        properties.setProperty(META_TABLE_COLUMNS, Joiner.on(',').join(handle.getColumnNames()));
+        Iterable<String> hiveTypeNames = transform(transform(dataColumnTypes, HiveType::toHiveType), HiveType::getHiveTypeName);
+
+        properties = new Properties();
+        properties.setProperty(META_TABLE_COLUMNS, Joiner.on(',').join(dataColumnNames));
         properties.setProperty(META_TABLE_COLUMN_TYPES, Joiner.on(':').join(hiveTypeNames));
 
-        serializer = initializeSerializer(conf, properties, handle.getHiveStorageFormat().getSerDe());
-        recordWriter = createRecordWriter(target, conf, properties, handle.getHiveStorageFormat().getOutputFormat());
+        if (serdeParameters != null) {
+            for (String key : serdeParameters.keySet()) {
+                properties.setProperty(key, serdeParameters.get(key));
+            }
+        }
+
+        try {
+            serializer = (Serializer) lookupDeserializer(serdeLib);
```

Comment:
Actually it is possible to specify different SerDe's for different partitions in Hive. Probably we should take it into account. Or at least we should throw an exception at the begging like "insert into partitioned tables with different serde settings per partition is not supported yet"